### DISCUSSION TITLE
About GraalVM support

### DISCUSSION THREAD

**Discussion by kijuky** [playframework/playframework#11253] [Ideas]
Hello everyone.

I'm interested in Play applications that run lightly in a cloud environment, and [GraalVM](https://www.graalvm.org/) seemed like a very good solution. I tried changing the [play-scala-seed](https://github.com/playframework/play-scala-seed.g8) project to GraalVM and the performance improved.

I want to make GraalVM easy to use with Play. I would like to improve Play with some fixes:

- GraalVM needs to support the `resource` scheme for external resource references ([implementation is the same as the existing` bundle` scheme](https://github.com/kijuky/playframework/compare/main...feature/add-resource-scheme))
- Finally, I want to output GraalVM compatible binaries with `.enablePlugin (PlayNativeSupport)` and `sbt graalvm-native-image: packageBin`.

I would love to hear your opinions.

# other reference

- Example of graalvm(JIT) on play https://discuss.lightbend.com/t/blog-post-running-play-on-graalvm/1005
- Example of graalvm support in spring boot https://docs.spring.io/spring-native/docs/current/reference/htmlsingle/

**Comment by francisdb** [comment]
Do you have an idea on the effort required to get to a native binary?

**Reply by kijuky** [reply]
Many are still under consideration, but some are mandatory.

# must

- apply `resource` scheme

# Not yet materialized

- Supply method about reflect-config.json and resource-config.json
- Creating an sbt plugin that summarizes the required settings
  - About `PlayKeys.externalizeResources := false` etc...

I'm experimenting with [this repository](https://github.com/kijuky/play-on-graalvm) (still in the experimental stage). Since it is necessary to support the `resource` scheme, I would like you to incorporate it into Play even during the experimental stage.

**Reply by kijuky** [reply]
And we need to have reflect-config.json as an important task. Akka already has [akka-graal-config](https://github.com/vmencik/akka-graal-config). I think we need to have a similar graal-XXXX project on Play as well.

First, I think we need to identify the libraries for which we need to create additional repositories(for graal-XXXX), based on [the sample project reflect-config.json](https://github.com/kijuky/play-on-graalvm/blob/main/reflect-config.json).

**Reply by kijuky** [reply]
Actually, I think it is realistic to divide it into several phases.

Phase 1

- Modify the `resource` scheme in Play.
- This allows the user to try to create a native image with the appropriate resource-config.json, reflect-config.json and sbt settings.

Phase 2

- Provides resource-config.json and reflect-config.json for Play and the libraries it depends on.
- This allows the user to try to create a native image with the appropriate sbt settings.

Phase 3

- Provides an sbt plugin that summarizes the required sbt settings.
- This allows users to try to create a native image by enabling the sbt plugin.

**Reply by jtjeferreira** [reply]
> Provides an sbt plugin that summarizes the required sbt settings

why not just use https://github.com/sbt/sbt-native-packager and https://sbt-native-packager.readthedocs.io/en/stable/formats/graalvm-native-image.html

**Reply by mkurz** [reply]
@kijuky Thank you very much for your interest and efforts. Since I never build native images with GraalVM myself yet, I lack a bit of knowledge of how things fit together. Would it possible to just reuse the sbt-native-packager GraalVM Native Image Plugin like mentioned by @jtjeferreira? If on top of that we still need to make modifications to Play I am happy to merge pull requests you provide (I guess we are fine merging the additional `resource` scheme, however meanwhile you can just publish Play locally to make that work for you).
Also if there is the need for another project like `akka-graal-config` (but for Play), I think the best solution is you put that under your own GitHub user account and we could transfer it to the Play organization later when its ready for prime time and if everything looks good. We of course could also add that project directly to the Play repo as well, the way we do that for all play modules. But we will figure that out.

**Reply by kijuky** [reply]
> why not just use https://github.com/sbt/sbt-native-packager and https://sbt-native-packager.readthedocs.io/en/stable/formats/graalvm-native-image.html

> Would it possible to just reuse the sbt-native-packager GraalVM Native Image Plugin like mentioned by @jtjeferreira? 

I'm sorry, the explanation wasn't enough.

Of course, use the sbt-native-package used for play. This plugin is great. The library I'd like to provide is to rely on sbt-native-package to _hide_ the parameters needed to create a native-image from the user's build.sbt. For example, `PlayKeys.externalizeResources: = false` or` "--enable-http" `.

> If on top of that we still need to make modifications to Play I am happy to merge pull requests you provide (I guess we are fine merging the additional resource scheme, however meanwhile you can just publish Play locally to make that work for you).

Thank you for your consideration. In fact, I built Play locally to run Play on GraalVM. My development experience was very good as a product as big as Play followed the sbt standard commands.

However, I want to make it easier to experiment with builds on GraalVM. Therefore, I would like to issue a pull request first only to modify the `resource` protocol.

> Also if there is the need for another project like akka-graal-config (but for Play), I think the best solution is you put that under your own GitHub user account and we could transfer it to the Play organization later when its ready for prime time and if everything looks good. We of course could also add that project directly to the Play repo as well, the way we do that for all play modules. But we will figure that out.

I think the preparation of reflect-config.json is a very interesting discussion. At the moment I also prefer something like akka-graal-config. As another implementation method, Spring creates reflect-config.json from the result of AOT. I think this idea is a smart way to be less prepared for users. I would like to find a "smart" solution in Play through this discussion with you.

**Reply by mkurz** [reply]
You might be interested in this file: https://github.com/playframework/playframework/blob/2.8.15/dev-mode/sbt-plugin/src/main/scala/play/sbt/Play.scala Here you can find all the sbt plugins Play provides by default. At the bottom you can see `PlayAkkaHttpServer` and `PlayNettyServer`. The source for those modules can be found in [`transport/server`](https://github.com/playframework/playframework/tree/main/transport/server). Maybe this would be a way to set things up for Graal Support as well?

**Reply by kijuky** [reply]
> You might be interested in this file: https://github.com/playframework/playframework/blob/2.8.15/dev-mode/sbt-plugin/src/main/scala/play/sbt/Play.scala Here you can find all the sbt plugins Play provides by default. At the bottom you can see PlayAkkaHttpServer and PlayNettyServer. The source for those modules can be found in [transport/server](https://github.com/playframework/playframework/tree/main/transport/server). Maybe this would be a way to set things up for Graal Support as well?

Oh, that's exactly the file I wanted to work on ...!

Yes, in the end I would like to provide the settings for creating a native image in the form of `.enablePlugins(PlayNative)` or `.enablePlugins(PlayScala, PlayNativeSupport)`.

**Comment by kijuky** [comment]
MEMO

- play has many pluggable parts and different DI methods. Graalvm support for them is very painstaking. At first, it seems better to consider support assuming a specific configuration.
- At first, I would like to check the operation in my personal repository, and if the problem is solved, I would like you to include that repository in playframework.
- The specific support contents are as follows:
  - Create reflect-config.json, resource-config.json
  - Implicit definition of build settings required for graalvm

**Comment by kijuky** [comment]
MEMO issues link

https://github.com/playframework/playframework/issues/9402
https://github.com/playframework/playframework/issues/11731
